File: python/diffcheck.py
"""
diffcheck.py — Compare two Word documents and produce a diff .docx.

Removed words appear in RED with STRIKETHROUGH.
Added words appear in GREEN.
Unchanged words appear in the default style.

Usage (called by Next.js API route):
    python3 diffcheck.py <original.docx> <revised.docx> <output.docx>

Usage (legacy / standalone):
    python3 diffcheck.py
    # compares SampleDoc.docx vs SampleDoc1.docx → SampleDoc_diff.docx
"""

import sys
import difflib
from docx import Document
from docx.shared import RGBColor


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def extract_paragraphs(doc_path: str) -> list:
    """Return a list of paragraphs; each paragraph is a list of word tokens."""
    doc = Document(doc_path)
    result = []
    for para in doc.paragraphs:
        text = para.text.strip()
        if text:
            result.append(text.split())
    return result


def diff_word_lists(old_words: list, new_words: list):
    """
    Run a SequenceMatcher diff on two word lists.
    Yields tuples of (tag, word) where tag is one of:
        'equal'   - unchanged
        'delete'  - in old only  (render red + strikethrough)
        'insert'  - in new only  (render green)
    """
    matcher = difflib.SequenceMatcher(None, old_words, new_words, autojunk=False)
    for opcode, i1, i2, j1, j2 in matcher.get_opcodes():
        if opcode == "equal":
            for w in old_words[i1:i2]:
                yield ("equal", w)
        elif opcode == "delete":
            for w in old_words[i1:i2]:
                yield ("delete", w)
        elif opcode == "insert":
            for w in new_words[j1:j2]:
                yield ("insert", w)
        elif opcode == "replace":
            for w in old_words[i1:i2]:
                yield ("delete", w)
            for w in new_words[j1:j2]:
                yield ("insert", w)


# ---------------------------------------------------------------------------
# Core
# ---------------------------------------------------------------------------

def generate_diff_doc(original_path: str, revised_path: str, output_path: str):
    orig_paras = extract_paragraphs(original_path)
    rev_paras  = extract_paragraphs(revised_path)

    # Align paragraphs by index; pad the shorter list with empty paragraphs
    max_len = max(len(orig_paras), len(rev_paras))
    orig_paras += [[]] * (max_len - len(orig_paras))
    rev_paras  += [[]] * (max_len - len(rev_paras))

    diff_doc = Document()

    stats = {"equal": 0, "delete": 0, "insert": 0}

    for old_words, new_words in zip(orig_paras, rev_paras):
        paragraph = diff_doc.add_paragraph()

        for tag, word in diff_word_lists(old_words, new_words):
            stats[tag] += 1
            run = paragraph.add_run(word + " ")

            if tag == "delete":
                run.font.color.rgb = RGBColor(192, 57, 43)   # red
                run.font.strike    = True
            elif tag == "insert":
                run.font.color.rgb = RGBColor(39, 174, 96)   # green
            # equal -> default formatting

    diff_doc.save(output_path)

    print(
        "[diffcheck] Done -- "
        + str(stats["equal"]) + " unchanged, "
        + str(stats["delete"]) + " removed, "
        + str(stats["insert"]) + " added -> " + output_path
    )

    # Return stats so the API route can surface them
    return stats


# ---------------------------------------------------------------------------
# Entry point
# ---------------------------------------------------------------------------

if __name__ == "__main__":
    args = sys.argv[1:]

    if len(args) == 3:
        original_path, revised_path, output_path = args
    elif len(args) == 0:
        original_path = "SampleDoc.docx"
        revised_path  = "SampleDoc1.docx"
        output_path   = "SampleDoc_diff.docx"
    else:
        print(
            "Usage: python3 diffcheck.py [original.docx revised.docx output.docx]",
            file=sys.stderr,
        )
        sys.exit(1)

    generate_diff_doc(original_path, revised_path, output_path)

File: lib/diffcheck.js
/**
 * diffcheck.js
 * Pure Node.js equivalent of diffcheck.py
 *
 * Reads two .docx files, diffs them word-by-word using the
 * Myers diff algorithm, and writes a new .docx where:
 *   - removed words are RED + strikethrough
 *   - added words are GREEN
 *   - unchanged words are unstyled
 *
 * Dependencies (already in Next.js ecosystem, just npm install):
 *   npm install docx mammoth
 *
 *   mammoth  — extracts plain text from .docx buffers
 *   docx     — builds the output .docx document
 */
import { diffWords as jsDiff } from "diff";
import mammoth from "mammoth";
import {
  Document,
  Packer,
  Paragraph,
  Run,
  TextRun,
} from "docx";

// ---------------------------------------------------------------------------
// 1. Extract text from a .docx buffer
//    Returns an array of paragraphs, each an array of word strings.
// ---------------------------------------------------------------------------
async function extractParagraphs(buffer) {
  const { value } = await mammoth.extractRawText({ buffer });
  return value
    .split(/\r?\n/)
    .map((line) => line.trim())
    .filter(Boolean)
    .map((line) => line.split(/\s+/).filter(Boolean));
}

// ---------------------------------------------------------------------------
// 2. Myers / SequenceMatcher-style diff on two word arrays
//    Returns an array of { tag, word } objects where tag is:
//      "equal"  — word is in both
//      "delete" — word only in old  → red + strikethrough
//      "insert" — word only in new  → green
// ---------------------------------------------------------------------------
function diffWords(oldWords, newWords) {
  const oldStr = oldWords.join(" ");
  const newStr = newWords.join(" ");

  const result = [];
  for (const part of jsDiff(oldStr, newStr)) {
    const words = part.value.trim().split(/\s+/).filter(Boolean);
    for (const word of words) {
      if (part.added)   result.push({ tag: "insert", word });
      else if (part.removed) result.push({ tag: "delete", word });
      else              result.push({ tag: "equal",  word });
    }
  }
  return result;
}

// ---------------------------------------------------------------------------
// 3. Build a .docx document from the diff and return it as a Buffer
// ---------------------------------------------------------------------------
async function buildDiffDoc(diffParagraphs) {
  const children = diffParagraphs.map((tokens) => {
    const runs = tokens.map(({ tag, word }) => {
      if (tag === "delete") {
        return new TextRun({
          text: word + " ",
          color: "C0392B",   // red
          strike: true,
        });
      }
      if (tag === "insert") {
        return new TextRun({
          text: word + " ",
          color: "27AE60",   // green
        });
      }
      // equal
      return new TextRun({ text: word + " " });
    });

    return new Paragraph({ children: runs });
  });

  const doc = new Document({ sections: [{ children }] });
  return Packer.toBuffer(doc);
}

// ---------------------------------------------------------------------------
// 4. Main exported function — mirrors generate_diff_doc() in diffcheck.py
//    Accepts Node.js Buffers, returns { docxBuffer, stats }
// ---------------------------------------------------------------------------
export async function generateDiffDoc(originalBuffer, revisedBuffer) {
  const [origParas, revParas] = await Promise.all([
    extractParagraphs(originalBuffer),
    extractParagraphs(revisedBuffer),
  ]);

  // Pad the shorter list so paragraph pairs always align
  const maxLen = Math.max(origParas.length, revParas.length);
  while (origParas.length < maxLen) origParas.push([]);
  while (revParas.length  < maxLen) revParas.push([]);

  const stats = { unchanged: 0, removed: 0, added: 0 };
  const diffParagraphs = [];

  for (let pi = 0; pi < maxLen; pi++) {
    const tokens = diffWords(origParas[pi], revParas[pi]);
    tokens.forEach(({ tag }) => {
      if (tag === "equal")  stats.unchanged++;
      if (tag === "delete") stats.removed++;
      if (tag === "insert") stats.added++;
    });
    diffParagraphs.push(tokens);
  }

  const docxBuffer = await buildDiffDoc(diffParagraphs);
  return { docxBuffer, stats };
}
// Add this export alongside generateDiffDoc

export async function getDiffParagraphs(originalBuffer, revisedBuffer) {
  const [origParas, revParas] = await Promise.all([
    extractParagraphs(originalBuffer),
    extractParagraphs(revisedBuffer),
  ]);

  const maxLen = Math.max(origParas.length, revParas.length);
  while (origParas.length < maxLen) origParas.push([]);
  while (revParas.length  < maxLen) revParas.push([]);

  const stats = { unchanged: 0, removed: 0, added: 0 };
  const diffParagraphs = [];

  for (let pi = 0; pi < maxLen; pi++) {
    const tokens = diffWords(origParas[pi], revParas[pi]);
    tokens.forEach(({ tag }) => {
      if (tag === "equal")  stats.unchanged++;
      if (tag === "delete") stats.removed++;
      if (tag === "insert") stats.added++;
    });
    diffParagraphs.push(tokens);
  }

  return { diffParagraphs, stats };
}